File: src/application/ports/paper_repository_port.py
"""
PaperRepositoryPort - Abstract interface for research paper data access.

This port defines the contract that concrete repository implementations must
follow. It represents the boundary between the application layer and the
infrastructure layer, following the Dependency Inversion Principle.

Educational Notes:
- This is a port (interface) in hexagonal architecture
- Defines WHAT the application needs, not HOW it's implemented
- Infrastructure layer will provide concrete implementations (adapters)
- Enables easy testing with mock implementations
- Follows Repository Pattern from Domain-Driven Design

Design Principles Applied:
- Dependency Inversion: Application depends on abstraction, not concretion
- Interface Segregation: Contains only methods needed by application
- Single Responsibility: Focused only on paper data access

Repository Pattern Benefits:
- Encapsulates data access logic
- Provides a consistent interface for different data sources
- Enables unit testing without database dependencies
- Supports multiple implementations (database, API, cache, etc.)
"""

from abc import ABC, abstractmethod
from typing import List, Optional

from domain.entities.research_paper import ResearchPaper
from domain.value_objects.search_query import SearchQuery


class PaperRepositoryPort(ABC):
    """
    Abstract repository interface for research paper data access.

    This port defines all the data access operations that the application
    layer needs to work with research papers. Concrete implementations
    will be provided by the infrastructure layer.

    Educational Note:
    Using ABC (Abstract Base Class) ensures that concrete implementations
    must provide all required methods. This enforces the contract at
    runtime and helps catch implementation errors early.

    Repository Pattern:
    - Provides a collection-like interface for domain objects
    - Encapsulates the logic needed to access data sources
    - Centralizes common data access functionality
    - Enables easier unit testing and maintainability
    """

    @abstractmethod
    def find_by_query(self, query: SearchQuery) -> List[ResearchPaper]:
        """
        Find research papers matching the given search query.

        This is the primary search method that takes a SearchQuery value object
        and returns a list of matching ResearchPaper entities. The concrete
        implementation will handle the actual search logic.

        Args:
            query: SearchQuery value object containing search parameters

        Returns:
            List[ResearchPaper]: List of papers matching the query.
                                Empty list if no matches found.

        Raises:
            RepositoryError: If search operation fails

        Educational Note:
        - Returns domain objects (ResearchPaper), not data structures
        - Takes domain objects (SearchQuery) as parameters
        - This keeps the repository focused on domain concepts
        - Error handling is implementation-specific but should be consistent
        """
        pass

    @abstractmethod
    def find_by_doi(self, doi: str) -> Optional[ResearchPaper]:
        """
        Find a research paper by its DOI (Digital Object Identifier).

        DOIs are unique identifiers for academic papers, so this method
        should return at most one paper.

        Args:
            doi: The DOI string to search for

        Returns:
            Optional[ResearchPaper]: The paper if found, None otherwise

        Raises:
            RepositoryError: If lookup operation fails

        Educational Note:
        DOI lookup is a common operation in academic systems, so we
        provide a dedicated method rather than forcing users to construct
        a SearchQuery for this simple case.
        """
        pass

    @abstractmethod
    def find_by_arxiv_id(self, arxiv_id: str) -> Optional[ResearchPaper]:
        """
        Find a research paper by its ArXiv identifier.

        ArXiv is a popular preprint server, especially for physics and
        computer science papers. ArXiv IDs are unique.

        Args:
            arxiv_id: The ArXiv ID string to search for

        Returns:
            Optional[ResearchPaper]: The paper if found, None otherwise

        Raises:
            RepositoryError: If lookup operation fails

        Educational Note:
        Like DOI lookup, this is a common enough operation to warrant
        its own method. It also demonstrates how the repository can
        support multiple identification schemes.
        """
        pass

    @abstractmethod
    def save_paper(self, paper: ResearchPaper) -> None:
        """
        Save a single research paper to the repository.

        This method handles both creating new papers and updating existing ones.
        The implementation should use the paper's identity (DOI or ArXiv ID)
        to determine whether this is a create or update operation.

        Args:
            paper: The ResearchPaper entity to save

        Raises:
            RepositoryError: If save operation fails
            ValidationError: If paper data is invalid

        Educational Note:
        - Repository handles the persistence details
        - Application doesn't need to know about create vs update logic
        - Domain object identity determines the operation type
        - This follows the Repository pattern's collection semantics
        """
        pass

    @abstractmethod
    def count_all(self) -> int:
        """
        Count total number of research papers in the repository.

        This method provides a quick way to get the total count of papers
        without retrieving all data. Useful for health checks, statistics,
        and pagination calculations.

        Returns:
            int: Total number of research papers stored

        Raises:
            RepositoryError: If count operation fails

        Educational Note:
        Count operations are useful for:
        - Health check endpoints to verify repository accessibility
        - Statistics and monitoring dashboards
        - Pagination calculations in UI layers
        - Performance monitoring of data growth
        """
        pass

    @abstractmethod
    def save_papers(self, papers: List[ResearchPaper]) -> None:
        """
        Save multiple research papers in a batch operation.

        This method should be more efficient than calling save_paper
        multiple times, especially for large batches. The implementation
        might use database transactions or batch APIs.

        Args:
            papers: List of ResearchPaper entities to save

        Raises:
            RepositoryError: If batch save operation fails
            ValidationError: If any paper data is invalid

        Educational Note:
        Batch operations are important for performance when dealing with
        large datasets. This method provides an optimization opportunity
        while maintaining the same semantics as individual saves.
        """
        pass


# Educational Notes for Students:
#
# 1. Port vs Adapter Pattern:
#    - Port: Interface defined by application (this file)
#    - Adapter: Concrete implementation in infrastructure layer
#    - This separation enables dependency inversion
#
# 2. Repository Pattern Benefits:
#    - Encapsulates data access complexity
#    - Provides domain-oriented interface
#    - Enables easy testing with mocks
#    - Supports multiple data sources
#
# 3. Abstract Base Classes:
#    - Enforce implementation contracts at runtime
#    - Provide clear documentation of required methods
#    - Enable isinstance() checks for type safety
#    - Python's way of defining interfaces
#
# 4. Domain-Focused Design:
#    - Methods take/return domain objects, not primitives
#    - Method names reflect business operations
#    - Error handling follows domain concepts
#    - Implementation details are hidden from application
#
# 5. Testing Strategy:
#    - Interface can be tested with mock implementations
#    - Tests define expected behavior for all implementations
#    - Concrete implementations tested separately
#    - Application layer tested using mocks of this interface

File: src/application/ports/PaperRepositoryPort.ts
import { Paper } from '../../domain/entities/Paper';

/**
 * PaperRepositoryPort - Abstract interface for paper data access operations.
 * 
 * This port demonstrates the Dependency Inversion Principle by defining the interface
 * that the application layer needs, without depending on specific implementations.
 * The infrastructure layer will provide concrete implementations of this interface.
 * 
 * Educational Notes:
 * - Shows Port pattern in Clean Architecture for external dependencies
 * - Demonstrates dependency inversion - application defines what it needs
 * - Enables testability through dependency injection and mocking
 * - Separates business logic from data access concerns
 * 
 * Design Decisions:
 * - Async methods support various data sources (files, APIs, databases)
 * - Promise-based API for consistent error handling
 * - Filter methods enable flexible paper retrieval strategies
 * - Repository pattern abstracts data access complexity
 * 
 * Real-World Application:
 * Academic research tools need flexible data access - papers might come from
 * ArXiv APIs, local files, databases, or web scraping. This interface allows
 * the application logic to remain unchanged regardless of data source.
 * 
 * Integration Points:
 * - Used by ExtractConceptsUseCase to retrieve papers for processing
 * - Implemented by infrastructure layer (ArxivPaperRepository, FilePaperRepository)
 * - Mocked in unit tests for isolated use case testing
 * - Supports future data source additions without application changes
 */

export interface PaperRepositoryPort {
    /**
     * Retrieve all available papers from the data source.
     * 
     * Educational Note:
     * This method demonstrates the Repository pattern's query interface.
     * The application layer calls this without knowing if papers come from
     * a database, API, or file system.
     * 
     * @returns Promise of all available Paper entities
     * @throws RepositoryError when data access fails
     */
    findAll(): Promise<Paper[]>;

    /**
     * Retrieve papers that match specific criteria.
     * 
     * Educational Note:
     * This method shows how repository interfaces can provide flexible
     * filtering while maintaining abstraction. The implementation details
     * of filtering (SQL queries, API parameters, etc.) are hidden.
     * 
     * @param criteria Filtering criteria for paper selection
     * @returns Promise of matching Paper entities
     * @throws RepositoryError when filtering fails
     */
    findByQuery(criteria: PaperQueryCriteria): Promise<Paper[]>;

    /**
     * Retrieve a specific paper by its unique identifier.
     * 
     * Educational Note:
     * Identity-based retrieval is fundamental to entity management.
     * This method supports use cases that need to process specific papers.
     * 
     * @param id Unique identifier for the paper
     * @returns Promise of Paper entity or undefined if not found
     * @throws RepositoryError when retrieval fails
     */
    findById(id: string): Promise<Paper | undefined>;

    /**
     * Retrieve papers that have sufficient text content for concept extraction.
     * 
     * Educational Note:
     * This method demonstrates domain-specific repository methods that
     * support particular use cases. The filtering logic understands
     * what constitutes "sufficient" content for concept extraction.
     * 
     * @returns Promise of Papers suitable for concept extraction
     * @throws RepositoryError when filtering fails
     */
    findWithSufficientContent(): Promise<Paper[]>;
}

/**
 * Criteria for querying papers from the repository.
 * 
 * Educational Note:
 * This interface demonstrates the Specification pattern by encapsulating
 * query criteria in a structured way. It provides type safety and
 * clear contracts for filtering operations.
 */
export interface PaperQueryCriteria {
    /** Search terms to match against paper content */
    searchTerms?: string[];
    
    /** Filter by publication date range */
    dateRange?: {
        startDate?: Date;
        endDate?: Date;
    };
    
    /** Filter by minimum abstract length for quality */
    minAbstractLength?: number;
    
    /** Filter by paper source (ArXiv, manual upload, etc.) */
    source?: string;
    
    /** Limit number of results for performance */
    limit?: number;
}

/**
 * Error thrown when repository operations fail.
 * 
 * Educational Note:
 * Domain-specific exceptions provide clear error boundaries and
 * enable appropriate error handling at each architectural layer.
 */
export class RepositoryError extends Error {
    constructor(message: string, public readonly cause?: Error) {
        super(message);
        this.name = 'RepositoryError';
    }
}
